contracts/andromeda_anchor/src/contract.rs

```diff
@@ -90,193 +86,187 @@ pub fn execute_deposit(
+    require(
+        info.funds.len() == 1,
+        ContractError::InvalidFunds {
+            msg: "Must deposit exactly 1 type of native coin.".to_string(),
+        },
+    )?;
+
     let config = CONFIG.load(deps.storage)?;
-    let coin_denom = config.stable_denom.clone();
-    let depositor = match recipient {
+    let recipient = match recipient {
         Some(recipient) => recipient,
         None => Recipient::Addr(info.sender.to_string()),
     };
 
+    let payment = &info.funds[0];
     require(
-        info.funds.len() <= 1usize,
-        ContractError::MoreThanOneCoin {},
+        payment.denom == UUSD_DENOM && payment.amount > Uint128::zero(),
+        ContractError::InvalidFunds {
+            msg: "Must deposit a non-zero quantity of uusd".to_string(),
+        },
     )?;
 
-    let payment = info
-        .funds
-        .iter()
-        .find(|x| x.denom == coin_denom && x.amount > Uint128::zero())
-        .ok_or_else(|| {
-            StdError::generic_err(format!("No {} assets are provided to deposit", coin_denom))
-        })?;
-    //create position
-    let position_idx = KEY_POSITION_IDX.load(deps.storage)?;
     let aust_balance = query_token_balance(
         &deps.querier,
-        deps.api.addr_humanize(&config.anchor_token)?,
+        deps.api.addr_humanize(&config.aust_token)?,
         env.contract.address,
     )?;
+    let recipient_addr = recipient.get_addr();
     PREV_AUST_BALANCE.save(deps.storage, &aust_balance)?;
+    RECIPIENT_ADDR.save(deps.storage, &recipient_addr)?;
     let payment_amount = payment.amount;
 
-    POSITION.save(
-        deps.storage,
-        &position_idx.u128().to_be_bytes(),
-        &Position {
-            idx: Default::default(),
-            owner: depositor,
-            deposit_amount: payment_amount,
-            aust_amount: Uint128::zero(),
-        },
-    )?;
+    if !POSITION.has(deps.storage, &recipient_addr) {
+        POSITION.save(
+            deps.storage,
+            &recipient_addr,
+            &Position {
+                recipient,
+                aust_amount: Uint128::zero(),
+            },
+        )?;
+    }
 
     //deposit Anchor Mint
     Ok(Response::new()
         .add_submessage(SubMsg::reply_on_success(
             CosmosMsg::Wasm(WasmMsg::Execute {
-                contract_addr: deps.api.addr_humanize(&config.anchor_mint)?.to_string(),
+                contract_addr: deps.api.addr_humanize(&config.anchor_market)?.to_string(),
                 msg: to_binary(&AnchorMarketMsg::DepositStable {})?,
                 funds: vec![payment.clone()],
             }),
-            1u64,
+            DEPOSIT_ID,
         ))
         .add_attributes(vec![
             attr("action", "deposit"),
             attr("deposit_amount", payment_amount),
         ]))
 }
 
-pub fn withdraw(
+pub fn execute_withdraw(
     deps: DepsMut,
     env: Env,
     info: MessageInfo,
-    position_idx: Uint128,
+    percent: Option<Uint128>,
```

Comment:
Why percentage over raw withdrawal amount?

Reply:
Just added a change that allows specifying either option.